3 course/OperatingSystem/lab2/sync_list_mutex.c

```diff
@@ -87,10 +91,38 @@ Storage* init(int count) {
 //как только перекинули указатель first->next на третий элемент, захватываем лочку на 2-4 элементах
 void random_nodes_swap(Storage* list) {
     assert(list->head);
-        
+    
+    pthread_mutex_lock(&list->sync_list);
     pthread_mutex_lock(&list->head->sync);
     Node* first = list->head;
+    
     Node* second = NULL;
+
+    if (first->next && rand() % 2 == 1) {
+        pthread_mutex_lock(&first->next->sync);
+        second = first->next;
+        if (second->next) {
+            pthread_mutex_lock(&second->next->sync);
+            first->next = second->next;
+            pthread_mutex_unlock(&second->next->sync);
+        } else {
+            first->next = second->next; //по факту будет нулл
+        }
+        second->next = first;
+        
+        list->head = second;
+        pthread_mutex_unlock(&list->sync_list);
+        pthread_mutex_unlock(&second->sync);
+
+    } else {
+        if (first->next) {
+            pthread_mutex_lock(&first->next->sync);
+            list->head = first->next;
```

Comment:
WTF is happening here ?  Please explain what you desired to have and what you actually have here.

Reply:
Writing code at night is a terrible idea... I thought it should be "first = first->next", but I wrote it incorrectly, and now I understand that this should not be, because we still change the elements of the second and third in the loop